File: leetcode48/Solution.java
class Solution {
    public void rotate(int[][] matrix) {
        int rows=matrix.length;
        int cols=matrix[0].length;
        
        // transpose the matrix
        for(int i=0; i<rows; i++) {
            for(int j=i+1; j<cols; j++) {
                int temp = matrix[i][j];
                matrix[i][j] = matrix[j][i];
                matrix[j][i]=temp;
            }
        }

        // reverse each rows in the matrix
        for(int i=0; i<rows; i++) {
            int L=0, R=rows-1;
            while(L<R) {
                int temp=matrix[i][L];
                matrix[i][L]=matrix[i][R];
                matrix[i][R]=temp;
                L++; R--;
            }
        }

        /**
        another way to reverse the matrix
        for(int i=0; i<n; i++) {
            for(int j=0; j<n/2; j++) {
                int temp = matrix[i][j]
                matrix[i][j] = matrix[i][n-j-1]
                matrix[i][n-j-1] = matrix[i][j]
            }
        }
         */
    }

  public void rotate(int[][] matrix) {
        int n=matrix.length;
        // flip the matrix top to bottom
        for(int i=0, j=n-1; i<j; j--,i++) {
            int[] temp = matrix[i];
            matrix[i] = matrix[j];
            matrix[j] = temp;
        }

        // find the transpose of the matrix
        for(int i=0; i<n; i++) {
            for(int j=i+1; j<n; j++) {
                int temp = matrix[i][j];
                matrix[i][j] = matrix[j][i];
                matrix[j][i] = temp;
            }
        }

    }

    public void rotate(int[][] mat) {
        int n = mat.length;
        for(int i = 0;i < (n+1)/2;i++) {
            for(int j = 0;j < n/2;j++) {
                int temp = mat[n-1-j][i];
                mat[n-1-j][i] = mat[n-1-i][n-1-j];
                mat[n-1-i][n-1-j] = mat[j][n-1-i];
                mat[j][n-1-i] = mat[i][j];
                mat[i][j] = temp;
            }
        }
    }
}

File: leetcode48/solution.py
class Solution:
    def rotate(self, matrix: List[List[int]]) -> None:
        """
        Do not return anything, modify matrix in-place instead.
        """
        n = len(matrix)
        for i in range(len(matrix)):
            for j in range(i, len(matrix)):
                matrix[i][j], matrix[j][i] = matrix[j][i], matrix[i][j]
        for row in matrix:
            row.reverse()

class Solution:
    def rotate(self, matrix:List[List[int]]) -> None:
        n=len(matrix)
        A=matrix

        # Find the transpose of the given matrix
        for i in range(n):
            for j in range(i+1, n):
                A[i][j], A[j][i] = A[j][i], A[i][j]

        # Find the reflection of the matrix
        for i in range(n):
            for j in range(n//2):
                A[i][j], A[i][n-j-1] = A[i][n-j-1], A[i][j]

class Solution:
    def rotate(self, matrix: List[List[int]]) -> None:
        Do not return anything, modify matrix in-place instead.
        n = len(matrix)
        for i in range((n+1)//2): # traverse the rows
            for j in range(n//2): # traverse the columns
                temp = matrix[n - 1 - j][i]
                matrix[n-1-j][i] = matrix[n-1-i][n-1-j]
                matrix[n-1-i][n-1-j] = matrix[j][n-1-i]
                matrix[j][n-1-i] = matrix[i][j]
                matrix[i][j] = temp

